File: JS/src/container_with_most_water.js
/**
 * Brute force approach (TLE)
 * @param {number[]} height
 * @return {number}
 */
function maxArea (height) {
  let max = 0

  for (let i = 0; i < height.length - 1; i++) {
    for (let j = i + 1; j < height.length; j++) {
      const area = Math.min(height[i], height[j]) * (j - i)
      max = Math.max(area, max)
    }
  }

  return max
}

/**
 * Two pointer approach
 * @param {number[]} height
 * @return {number}
 */
function maxArea2 (height) {
  let max = 0
  let left = 0
  let right = height.length - 1

  while (left < right) {
    const area = Math.min(height[left], height[right]) * (right - left)
    max = Math.max(area, max)

    if (height[left] === height[right]) {
      left++
      right--
    } else if (height[left] < height[right]) {
      left++
    } else {
      right--
    }
  }

  return max
}

export { maxArea, maxArea2 }

File: C++/src/container_with_most_water.cc
#include "container_with_most_water.hpp"

int Solution::maxArea(vector<int> &height) {
  int res = 0;

  int left = 0;
  int right = height.size() - 1;

  while (left < right) {
    int area = (right - left) * min(height[left], height[right]);
    res = max(res, area);

    if (height[left] == height[right]) {
      left++;
      right--;
    } else if (height[left] < height[right]) {
      left++;
    } else {
      right--;
    }
  }

  return res;
}
